File: 368. Largest Divisible Subset/main.py
from typing import List


class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        nums.sort()
        cache = {}

        def dfs(i):
            if i == len(nums): return []
            if i in cache: return cache[i]
            res = [nums[i]]
            for j in range(i + 1, len(nums)):
                if nums[j] % nums[i] == 0:
                    tmp = [nums[i]] + dfs(j)
                    res = tmp if len(tmp) > len(res) else res
            cache[i] = res
            return res
        
        res = []
        for i in range(len(nums)):
            tmp = dfs(i)
            res = tmp if len(tmp) > len(res) else res
        return res
                

File: 368. Largest Divisible Subset/main.go
package main

import "slices"

func largestDivisibleSubset(nums []int) []int {
	slices.Sort(nums)
	cache := make(map[int][]int)

	var dfs func(i int) []int
	dfs = func(i int) []int {
		if i == len(nums) {
			return []int{}
		}
		if cached, exists := cache[i]; exists {
			return cached
		}
		res := []int{nums[i]}
		for j := i + 1; j < len(nums); j++ {
			if nums[j]%nums[i] == 0 {
				tmp := append([]int{nums[i]}, dfs(j)...)
				if len(tmp) > len(res) {
					res = tmp
				}
			}
		}
		cache[i] = res
		return res
	}

	res := []int{}
	for i := range len(nums) {
		tmp := dfs(i)
		if len(tmp) > len(res) {
			res = tmp
		}
	}
	return res
}
